File: server/routes/code.py
"""
Code execution routes with run-and-save functionality
"""
from fastapi import APIRouter, Depends, Header
from typing import Optional
from models import (
    CodeRunRequest, CodeRunResponse,
    CodeRunAndSaveRequest, CodeRunAndSaveResponse
)
from services.code_service import run_code
from services.hint_service import generate_hints
from services.submissions_service import create_submission
from routes.auth import get_current_user, get_optional_user

router = APIRouter()


@router.post("/run", response_model=CodeRunResponse)
async def execute_code(request: CodeRunRequest):
    """Execute code in sandbox and return results (no auth required)"""
    result = run_code(
        code=request.code,
        language=request.language,
        user_input=request.input or ""
    )
    return CodeRunResponse(
        success=result["success"],
        output=result["output"],
        compilation_result=result["compilation_result"],
        execution_time=result["execution_time"],
        status=result["status"]
    )


@router.post("/run-and-save", response_model=CodeRunAndSaveResponse)
async def execute_and_save(
    request: CodeRunAndSaveRequest,
    user: dict = Depends(get_current_user)
):
    """
    Execute code, generate hints if error, and save to user's history.
    Requires authentication.
    """
    # Run the code
    result = run_code(
        code=request.code,
        language=request.language,
        user_input=request.input or ""
    )
    
    # Generate hints if there's an error and hints are requested
    hints_data = None
    if not result["success"] and request.get_hints:
        error_msg = result.get("compilation_result") or result.get("output") or ""
        hints_data = generate_hints(
            code=request.code,
            language=request.language,
            error=error_msg,
            expected_output=request.expected_output or ""
        )
    
    # Determine status for submission
    status = "success" if result["success"] else "error"
    error_type = hints_data.get("error_type") if hints_data else None
    hints_list = hints_data.get("hints") if hints_data else None
    root_cause = hints_data.get("root_cause") if hints_data else None
    
    # Save submission to history
    submission = create_submission(
        user_id=user["id"],
        code=request.code,
        language=request.language,
        output=result.get("output", ""),
        status=status,
        execution_time=result.get("execution_time", 0),
        error_type=error_type,
        hints=hints_list,
        root_cause=root_cause
    )
    
    return CodeRunAndSaveResponse(
        success=result["success"],
        output=result["output"],
        compilation_result=result["compilation_result"],
        execution_time=result["execution_time"],
        status=result["status"],
        submission_id=submission["id"],
        error_type=error_type,
        hints=hints_list,
        root_cause=root_cause
    )


@router.post("/debug", response_model=CodeRunAndSaveResponse)
async def debug_code(
    request: CodeRunAndSaveRequest,
    user: Optional[dict] = Depends(get_optional_user)
):
    """
    Run code with debugging - always generate hints on errors.
    Works with or without authentication.
    If authenticated, saves to history.
    """
    # Run the code
    result = run_code(
        code=request.code,
        language=request.language,
        user_input=request.input or ""
    )
    
    # Generate hints if there's an error
    hints_data = None
    if not result["success"]:
        error_msg = result.get("compilation_result") or result.get("output") or ""
        hints_data = generate_hints(
            code=request.code,
            language=request.language,
            error=error_msg,
            expected_output=request.expected_output or ""
        )
    
    # Prepare response data
    status = "success" if result["success"] else "error"
    error_type = hints_data.get("error_type") if hints_data else None
    hints_list = hints_data.get("hints") if hints_data else None
    root_cause = hints_data.get("root_cause") if hints_data else None
    submission_id = None
    
    # Save to history if authenticated
    if user:
        submission = create_submission(
            user_id=user["id"],
            code=request.code,
            language=request.language,
            output=result.get("output", ""),
            status=status,
            execution_time=result.get("execution_time", 0),
            error_type=error_type,
            hints=hints_list,
            root_cause=root_cause
        )
        submission_id = submission["id"]
    
    return CodeRunAndSaveResponse(
        success=result["success"],
        output=result["output"],
        compilation_result=result["compilation_result"],
        execution_time=result["execution_time"],
        status=result["status"],
        submission_id=submission_id,
        error_type=error_type,
        hints=hints_list,
        root_cause=root_cause
    )

File: server/src/routes/code.ts
import { Router } from 'express';
import { codeService } from '../services/codeService';
import { hintService } from '../services/hintService';
import { submissionsService } from '../services/submissionsService';
import { authMiddleware, optionalAuthMiddleware } from '../middleware/auth';
import { CodeRunRequest, CodeRunAndSaveRequest } from '../types';

const router = Router();

// Run code (no auth)
router.post('/run', async (req, res) => {
    try {
        const { code, language, input }: CodeRunRequest = req.body;
        const result = await codeService.runCode(code, language || 'python', input || '');
        res.json(result);
    } catch (error: any) {
        res.status(500).json({ detail: error.message || 'Code execution failed' });
    }
});

// Run and save (auth required)
router.post('/run-and-save', authMiddleware, async (req: any, res) => {
    try {
        const { code, language, input, expected_output, get_hints }: CodeRunAndSaveRequest = req.body;

        // Run code
        const result = await codeService.runCode(code, language || 'python', input || '');

        // Generate hints if error
        let hintsData = null;
        if (!result.success && (get_hints !== false)) {
            const errorMsg = result.compilation_result || result.output || '';
            hintsData = await hintService.generateHints(
                code,
                language || 'python',
                errorMsg,
                expected_output || ''
            );
        }

        // Save submission
        const submission = submissionsService.createSubmission(req.user.id, {
            code,
            language: language || 'python',
            output: result.output,
            status: result.success ? 'success' : 'error',
            execution_time: result.execution_time,
            error_type: hintsData?.error_type,
            hints: hintsData?.hints,
            root_cause: hintsData?.root_cause,
        });

        res.json({
            ...result,
            submission_id: submission.id,
            error_type: hintsData?.error_type,
            hints: hintsData?.hints,
            root_cause: hintsData?.root_cause,
        });
    } catch (error: any) {
        res.status(500).json({ detail: error.message || 'Failed to run and save' });
    }
});

// Debug code (optional auth)
router.post('/debug', optionalAuthMiddleware, async (req: any, res) => {
    try {
        const { code, language, input, expected_output }: CodeRunAndSaveRequest = req.body;

        // Run code
        const result = await codeService.runCode(code, language || 'python', input || '');

        // Generate hints if error
        let hintsData = null;
        if (!result.success) {
            const errorMsg = result.compilation_result || result.output || '';
            hintsData = await hintService.generateHints(
                code,
                language || 'python',
                errorMsg,
                expected_output || ''
            );
        }

        // Save if authenticated
        let submissionId;
        if (req.user) {
            const submission = submissionsService.createSubmission(req.user.id, {
                code,
                language: language || 'python',
                output: result.output,
                status: result.success ? 'success' : 'error',
                execution_time: result.execution_time,
                error_type: hintsData?.error_type,
                hints: hintsData?.hints,
                root_cause: hintsData?.root_cause,
            });
            submissionId = submission.id;
        }

        res.json({
            ...result,
            submission_id: submissionId,
            error_type: hintsData?.error_type,
            hints: hintsData?.hints,
            root_cause: hintsData?.root_cause,
        });
    } catch (error: any) {
        res.status(500).json({ detail: error.message || 'Debug failed' });
    }
});

export default router;
